File: Aryan's Satellite Control/claude.ai.java
import java.io.*;
import java.util.*;

public class Main {
    static class FenwickTree {
        private final int[] tree;

        FenwickTree(int n) {
            tree = new int[n + 1];
        }

        void update(int index, int delta) {
            for (index++; index < tree.length; index += index & -index)
                tree[index] += delta;
        }

        int query(int index) {
            int result = 0;
            for (index++; index > 0; index -= index & -index)
                result += tree[index];
            return result;
        }

        int findKth(int k) {
            int index = 0;
            int step = Integer.highestOneBit(tree.length - 1);

            while (step != 0) {
                int next = index + step;
                if (next < tree.length && tree[next] < k) {
                    index = next;
                    k -= tree[next];
                }
                step >>= 1;
            }

            return index;
        }
    }

    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner();
        StringBuilder output = new StringBuilder();

        int testCases = fs.nextInt();

        while (testCases-- > 0) {
            int n = fs.nextInt();
            int k = fs.nextInt();

            long[][] intervals = new long[n][2];
            long[] coordinates = new long[n + 1];
            coordinates[0] = 0;

            for (int i = 0; i < n; i++) {
                intervals[i][0] = fs.nextLong();
                intervals[i][1] = fs.nextLong();
                coordinates[i + 1] = intervals[i][1];
            }

            Arrays.sort(intervals, Comparator.comparingLong(a -> a[1]));
            Arrays.sort(coordinates);

            long[] values = new long[n + 1];
            int count = 0;

            for (long value : coordinates) {
                if (count == 0 || values[count - 1] != value)
                    values[count++] = value;
            }

            FenwickTree tree = new FenwickTree(count);
            tree.update(lowerBound(values, count, 0), k);

            int answer = 0;

            for (long[] interval : intervals) {
                int last = upperBound(values, count, interval[0]) - 1;

                if (last < 0)
                    continue;

                int available = tree.query(last);
                if (available == 0)
                    continue;

                int selected = tree.findKth(available);
                tree.update(selected, -1);
                tree.update(lowerBound(values, count, interval[1]), 1);
                answer++;
            }

            output.append(answer).append('\n');
        }

        System.out.print(output);
    }

    static int lowerBound(long[] array, int size, long value) {
        int left = 0, right = size;
        while (left < right) {
            int mid = (left + right) >>> 1;
            if (array[mid] < value) left = mid + 1;
            else right = mid;
        }
        return left;
    }

    static int upperBound(long[] array, int size, long value) {
        int left = 0, right = size;
        while (left < right) {
            int mid = (left + right) >>> 1;
            if (array[mid] <= value) left = mid + 1;
            else right = mid;
        }
        return left;
    }

    static class FastScanner {
        private final InputStream input = System.in;
        private final byte[] buffer = new byte[1 << 16];
        private int pointer = 0;
        private int length = 0;

        private int read() throws IOException {
            if (pointer >= length) {
                length = input.read(buffer);
                pointer = 0;
                if (length == -1) return -1;
            }
            return buffer[pointer++];
        }

        long nextLong() throws IOException {
            int c;
            do c = read(); while (c <= 32);
            long value = 0;
            while (c > 32) {
                value = value * 10 + c - '0';
                c = read();
            }
            return value;
        }

        int nextInt() throws IOException {
            return (int) nextLong();
        }
    }
}

File: Aryan's Satellite Control/claude.ai.cpp
#include <bits/stdc++.h>
using namespace std;

struct FenwickTree {
    int size;
    vector<int> tree;

    explicit FenwickTree(int n) : size(n), tree(n + 1, 0) {}

    void update(int index, int delta) {
        for (i++ndex; index <= size; index += index & -index)
            tree[index] += delta;
    }

    int query(int index) const {
        int result = 0;
        for (i++ndex; index > 0; index -= index & -index)
            result += tree[index];
        return result;
    }

    int findByOrder(int order) const {
        int index = 0;
        int step = 1;
        while ((step << 1) <= size) step <<= 1;

        for (; step; step >>= 1) {
            int next = index + step;
            if (next <= size && tree[next] < order) {
                index = next;
                order -= tree[next];
            }
        }

        return index;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--) {
        int N, K;
        cin >> N >> K;

        vector<pair<long long, long long>> intervals(N);
        vector<long long> ends(N);

        for (int i = 0; i < N; i++) {
            cin >> intervals[i].first >> intervals[i].second;
            ends[i] = intervals[i].second;
        }

        sort(intervals.begin(), intervals.end(),
             [](const auto& a, const auto& b) {
                 if (a.second != b.second) return a.second < b.second;
                 return a.first < b.first;
             });

        ends.push_back(0);
        sort(ends.begin(), ends.end());
        ends.erase(unique(ends.begin(), ends.end()), ends.end());

        FenwickTree fw((int)ends.size());
        int zeroIndex = lower_bound(ends.begin(), ends.end(), 0) - ends.begin();

        if (zeroIndex < (int)ends.size())
            fw.update(zeroIndex, K);

        int answer = 0;

        for (const auto& [left, right] : intervals) {
            int last = upper_bound(ends.begin(), ends.end(), left) - ends.begin() - 1;
            if (last < 0) continue;

            int count = fw.query(last);
            if (count == 0) continue;

            int chosen = fw.findByOrder(count);
            fw.update(chosen, -1);

            int finishIndex = lower_bound(ends.begin(), ends.end(), right) - ends.begin();
            fw.update(finishIndex, 1);

            ++answer;
        }

        cout << answer << '\n';
    }
}
